### PR TITLE
[TRTLLM-6741] [feat] enable LM tp for MTP, under attention dp case (cherry-pick #7128)

### PR DESCRIPTION
**Pull request by kaiyux**
## Summary by CodeRabbit

- New Features
  - Added optional LM head tensor-parallelism within Attention DP, including speculative decoding support. Enable via the new beta flag enable_lm_head_tp_in_adp.
  - Introduced LM_HEAD_TP_SIZE environment variable to control LM head TP group size.

- Performance
  - Improved scalability and throughput on multi-GPU setups when LM head TP in ADP is enabled.

- API
  - New public argument enable_lm_head_tp_in_adp (beta) added and included in serialized configs.

- Tests
  - Updated API stability references to include the new argument.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/llmapi/llm_args.py [hunk 1]
```diff
@@ -1261,6 +1263,11 @@ class BaseLlmArgs(StrictBaseModel):
         description="Enable attention data parallel.",
         status="beta")
 
+    enable_lm_tp_in_adp: bool = Field(
+        default=False,
+        description="Enable lm tp in attention dp.",
```
Comment (Superjomn) [inline tensorrt_llm/llmapi/llm_args.py:1268]: Should we capitalize "LM TP"?

**Review by Superjomn — APPROVED** [review]
LGTM on the llmapi changes

File: tensorrt_llm/llmapi/llm_args.py (hunk 1 above)
Comment (kaiyux) [reply tensorrt_llm/llmapi/llm_args.py]: Addressed. Thanks.

**Review by QiJune — APPROVED** [review]
LGTM

File: tensorrt_llm/_torch/models/modeling_deepseekv3.py [hunk 2]
```diff
@@ -167,10 +168,21 @@ def forward(self,
...
                 hidden_states = hidden_states[-1].unsqueeze(0)
 
-        if not (self.model_config.mapping.enable_attention_dp):
+        enable_attention_dp = self.model_config.mapping.enable_attention_dp
+        enable_lm_head_tp_in_adp = self.model_config.mapping.enable_lm_head_tp_in_adp
+
+        # Add pre-lm gather logic
+        if enable_lm_head_tp_in_adp:
+            # ADP + LM TP mode: perform All-Gather before LM_head
+            mapping_lm_head_tp = create_lm_head_tp_mapping(
```
Comment (lfr-0531) [inline tensorrt_llm/_torch/models/modeling_deepseekv3.py:177]: It would be better to move this `create_lm_head_tp_mapping` to `__init__`, not in the `forward`.

File: tensorrt_llm/_torch/modules/embedding.py [hunk 3]
```diff
@@ -80,12 +84,26 @@ def vocab_size_padded(self) -> int:
...
     def forward(
-            self,
-            input: torch.Tensor,
-            *,
-            all_reduce_params: Optional[AllReduceParams] = None
+        self,
+        input: torch.Tensor,
+        *,
+        all_reduce_params: Optional[AllReduceParams] = None,
+        is_mtp_head: bool = False,
```
Comment (lfr-0531) [inline tensorrt_llm/_torch/modules/embedding.py:91]: Can we change it to use `is_spec_decoding_head`? Eagle3 one-model can also use this feature.

File: tensorrt_llm/_torch/modules/embedding.py [hunk 4]
```diff
@@ -80,12 +84,26 @@ def vocab_size_padded(self) -> int:
...
+                self.mapping, 'enable_lm_head_tp_in_adp', False)):
+            tp_rank = self.mapping.tp_rank
+            tp_size = self.mapping.tp_size
+            tensor_shape = self.weight.shape
+            width = tensor_shape[0]
+            slice_width = math.ceil(width / tp_size)
+            slice_start = tp_rank * slice_width
+            slice_end = min((tp_rank + 1) * slice_width, width)
+            slice_obj = [slice(None)] * len(tensor_shape)
+            slice_obj[0] = slice(slice_start, slice_end)
```
Comment (lfr-0531) [inline tensorrt_llm/_torch/modules/embedding.py:103]: We can get this `slice_obj` in `__init__`. And for the `width`, plz use `self.out_features`.

File: tensorrt_llm/_torch/speculative/mtp.py [hunk 5]
```diff
@@ -473,9 +475,23 @@ def forward(
         for _, mtp_layer in enumerate(draft_model.mtp_layers):
             hidden_states = mtp_layer(embed_tokens=draft_model.embed_tokens,
                                       **draft_inputs)
-            logits = mtp_layer.shared_head(hidden_states, draft_model.lm_head,
+            token_count = hidden_states.view(-1,
+                                             hidden_states.shape[-1]).shape[0]
+            all_rank_max_num_tokens = attn_metadata.all_rank_max_num_tokens
+            pad_len = all_rank_max_num_tokens - token_count
+            if pad_len > 0:
```
Comment (lfr-0531) [inline tensorrt_llm/_torch/speculative/mtp.py:482]: Why do we need to pad the hidden states? And I'm not sure if CUDA graphs work with this "if pad_len > 0".

File: tensorrt_llm/_torch/speculative/mtp.py [hunk 6]
```diff
@@ -1095,6 +1112,20 @@ def draft_sampler(
             combined = self.get_local_max_and_combined(logits)
             gathered = allgather(combined, self.model_config.mapping, dim=-1)
             draft_tokens = self.get_draft_tokens_from_gathered(gathered)
+        elif (self.model_config is not None
+              and hasattr(self.model_config, 'mapping')
+              and self.model_config.mapping.tp_size
+              > 1) and self.model_config.mapping.enable_lm_head_tp_in_adp:
+            # For ADP + LM head TP mode, we need to find the global argmax across all TP ranks
+            mapping_lm_tp = create_lm_head_tp_mapping(self.model_config.mapping)
```
Comment (lfr-0531) [inline tensorrt_llm/_torch/speculative/mtp.py:1120]: Is it possible to just create one `mapping_lm_tp` in `modeling_deepseekv3`, then pass it to `MTPWork`, `MTPEagleWorker`, and `DeepseekV3MTPHead`? So we don't need to create it multiple times.

File: tensorrt_llm/_torch/speculative/mtp.py [hunk 7]
```diff
@@ -1194,10 +1225,27 @@ def prepare_position_ids_and_last_tokens(position_ids, attn_metadata):
                     **inputs)
                 # All of the seq_len are 1, use batch_indices_cuda as gather_ids
                 gather_ids = spec_metadata.batch_indices_cuda[:batch_size]
+            hidden_states_gathered = hidden_states[gather_ids]
+            token_count = hidden_states_gathered.view(
+                -1, hidden_states_gathered.shape[-1]).shape[0]
+            max_num_requests = spec_metadata.max_num_requests
+            pad_len = max_num_requests - token_count
+            if pad_len > 0:
```
Comment (lfr-0531) [inline tensorrt_llm/_torch/speculative/mtp.py:1233]: And also for this padding.

File: tensorrt_llm/_torch/modules/embedding.py [hunk 8]
```diff
@@ -80,12 +84,26 @@ def vocab_size_padded(self) -> int:
...
-            *,
-            all_reduce_params: Optional[AllReduceParams] = None
+        self,
+        input: torch.Tensor,
+        *,
+        all_reduce_params: Optional[AllReduceParams] = None,
+        is_mtp_head: bool = False,
     ) -> torch.Tensor:
-        output = super().forward(input, all_reduce_params=all_reduce_params)
+        if is_mtp_head and (self.mapping.enable_attention_dp and getattr(
```
Comment (yweng0828) [inline tensorrt_llm/_torch/modules/embedding.py:93]: Can we move these calculations to somewhere else like `__init__` ? These values ​​are unchangeable during the model run, is it right?

**Review by yweng0828 — APPROVED** [review]

File: tensorrt_llm/_torch/speculative/mtp.py (hunk 5 above)
Comment (Njuapp) [reply tensorrt_llm/_torch/speculative/mtp.py:482]: This is because nccl all gather needs batch size of all ranks to be equal, so we have to pad before all-gather.

Comment (Njuapp) [reply tensorrt_llm/_torch/speculative/mtp.py:482]: CUDA graph works if pad_len > 0.

File: tensorrt_llm/_torch/models/modeling_deepseekv3.py [hunk 9]
```diff
@@ -167,10 +170,19 @@ def forward(self,
             else:
                 hidden_states = hidden_states[-1].unsqueeze(0)
 
-        if not (self.model_config.mapping.enable_attention_dp):
```
Comment (litaotju) [inline tensorrt_llm/_torch/models/modeling_deepseekv3.py:170]: Thanks for the change.
Lets hold this PR , not merge to this SA branch, and only target to main branch. 
Lets stablize the branch. Thx

**Review by litaotju — CHANGES_REQUESTED** [review]

File: tensorrt_llm/_torch/speculative/mtp.py (hunk 5 above)
Comment (lfr-0531) [reply tensorrt_llm/_torch/speculative/mtp.py:481]: Discussed with @Njuapp offline, he will use the `all_rank_num_tokens` in spec_metadata for allgather and remove this padding. 

Currently, the MTP vanilla cannot support attention DP, so we will focus on MTP Eagle first.

File: tensorrt_llm/_torch/models/modeling_deepseekv3.py (hunk 9 above)
Comment (litaotju) [reply tensorrt_llm/_torch/models/modeling_deepseekv3.py:170]: Yes,target to main as we offline discussed.

**Review by nv-guomingz — APPROVED** [review]
LGTM on llm api part.

**Review by lfr-0531 — APPROVED** [review]

**Review by litaotju — APPROVED** [review]

File: tensorrt_llm/llmapi/llm_args.py [hunk 10]
```diff
@@ -1264,6 +1266,11 @@ class BaseLlmArgs(StrictBaseModel):
         description="Enable attention data parallel.",
         status="beta")
 
+    enable_lm_head_tp_in_adp: bool = Field(
+        default=False,
+        description="Enable LM head TP in attention dp.",
+        status="beta")
```
[the remark this answers is not in the document]
Comment (Njuapp) [reply tensorrt_llm/llmapi/llm_args.py:1272]: Yes, I submitted a new PR for this line of code: https://github.com/NVIDIA/TensorRT-LLM/pull/7891
